### PR TITLE
[pigeon] cleaned up run_test.sh

### PR DESCRIPTION
**Pull request by gaaclarke**
changes:
* added ability to specify what class of test you want to run
* added list of available tests
* stages are all refactored into functions
* dart compilation tests run on null-safe and non-null-safe now (in parallel)
* removed `-x` added some print statements
* moved to bash
* made variables all lowercase to avoid collisions

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated `CHANGELOG.md` to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the `#hackers-new` channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/pigeon/run_tests.sh [hunk 1]
```diff
@@ -86,195 +90,337 @@ test_pigeon_android() {
...
+  pub run pigeon \
+    --no-dart_null_safety \
+    --input $1 \
+    --dart_out $temp_dir_2/pigeon.dart &
+  non_null_safe_gen_pid=$!
+
+  wait $null_safe_gen_pid
+  wait $non_null_safe_gen_pid
+
+  dartanalyzer $temp_dir_1/pigeon.dart --fatal-infos --fatal-warnings --packages ./e2e_tests/test_objc/.packages &
```
Comment (xster) [inline packages/pigeon/run_tests.sh:132]: what's ./e2e_tests/test_objc/.packages about? Add a comment?

File: packages/pigeon/run_tests.sh [hunk 2]
```diff
@@ -86,195 +90,337 @@ test_pigeon_android() {
...
-cd mock_handler_tester
-flutter test
-popd
+run_mock_handler_tests() {
+  pushd $PWD
+  pub run pigeon \
+    --input pigeons/message.dart \
+    --dart_out mock_handler_tester/test/message.dart \
+    --dart_test_out mock_handler_tester/test/test.dart
+  dartfmt -w mock_handler_tester/test/message.dart
```
Comment (xster) [inline packages/pigeon/run_tests.sh:207]: why the dartfmt here?

File: packages/pigeon/run_tests.sh [hunk 3]
```diff
@@ -86,195 +90,337 @@ test_pigeon_android() {
...
-test_pigeon_ios ./pigeons/all_datatypes.dart
+run_dart_compilation_tests() {
+  # Make sure the artifacts are present.
+  flutter precache
+  # Make sure flutter dependencies are available.
+  pushd $PWD
+  cd e2e_tests/test_objc/
+  flutter pub get
+  popd
+  test_pigeon_dart ./pigeons/all_datatypes.dart
```
Comment (xster) [inline packages/pigeon/run_tests.sh:222]: do we also need to add a warning somewhere saying if you add more tests, you must list them here?

File: packages/pigeon/run_tests.sh [hunk 4]
```diff
@@ -86,195 +90,337 @@ test_pigeon_android() {
...
-cd platform_tests/ios_unit_tests
-flutter build ios --simulator
-cd ios
-xcodebuild \
+run_java_compilation_tests() {
+  # Make sure the artifacts are present.
+  flutter precache
+  # Make sure flutter dependencies are available.
+  pushd $PWD
+  cd e2e_tests/test_objc/
```
Comment (xster) [inline packages/pigeon/run_tests.sh:238]: what's the significance of e2e_tests/test_objc/ in the java test? If there's a special pigeon schema file in that test, should it be moved somewhere more generic?

File: packages/pigeon/run_tests.sh [hunk 5]
```diff
@@ -86,195 +90,337 @@ test_pigeon_android() {
...
+    *)
+      echo "unrecognized test: $OPTARG"
+      exit 1
+      ;;
+    esac
+    ;;
+  l)
+    echo "available tests for -t:
+  dart_compilation_tests - Compilation tests on generated Dart code.
+  dart_unittests         - Unit tests on and analysis on Pigeon's Dart source code.
```
Comment (xster) [inline packages/pigeon/run_tests.sh:370]: seems like this one test is different than the rest? Should it be called something more special? 

Not clear to me whether this is a unit test against the pigeon tool implementation itself (i.e. unit tests the ast.dart etc) or it's a unit test against the pigeon schema file.

**Review by xster — APPROVED** [review]

File: packages/pigeon/run_tests.sh (hunk 2 above)
Comment (gaaclarke) [reply packages/pigeon/run_tests.sh:207]: Those files can be checked in.  I'm still on the fence of if thats a good idea.  In my PR's I never check them in, but external contributors always do and it makes it easier to review their changes.

File: packages/pigeon/run_tests.sh (hunk 3 above)
Comment (gaaclarke) [reply packages/pigeon/run_tests.sh:222]: I don't think so.  When doing TDD you'll add the tests here as you implement them for the different generators.

File: packages/pigeon/run_tests.sh (hunk 5 above)
Comment (gaaclarke) [reply packages/pigeon/run_tests.sh]: reworded

File: packages/pigeon/run_tests.sh (hunk 4 above)
Comment (gaaclarke) [reply packages/pigeon/run_tests.sh:241]: It's for the aforementioned hack to get access to Flutter.   Added comment

File: packages/pigeon/run_tests.sh (hunk 1 above)
Comment (gaaclarke) [reply packages/pigeon/run_tests.sh:134]: Commented, I could probably make flutter a dev dependency for pigeon as an alternative.

**Pull request merged**